x/datapool/keeper/pool.go

```diff
@@ -553,3 +544,129 @@ func (k Keeper) increaseNumIssuedNFT(ctx sdk.Context, pool *types.Pool) {
+	pool, err := k.GetPool(ctx, redeemNFT.PoolId)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if pool.GetRound() != redeemNFT.Round {
+		return nil, types.ErrRoundNotMatched
+	}
+
+	nftContractAcc, err := sdk.AccAddressFromBech32(pool.NftContractAddr)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress
+	}
+
+	moduleAddr := types.GetModuleAddress()
+	transferNFTMsg := types.NewTransferNFTMsg(moduleAddr.String(), strconv.FormatUint(redeemNFT.NftId, 10))
+
+	transferNFTMsgBz, err := json.Marshal(transferNFTMsg)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemerAcc, err := sdk.AccAddressFromBech32(redeemNFT.Redeemer)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemTokenId, err := k.GetRedeemerDataPass(ctx, pool.PoolId, redeemerAcc)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if !contains(redeemTokenId, strconv.FormatUint(redeemNFT.NftId, 10)) {
+		return nil, types.ErrInvalidTokenId
+	}
+
+	zeroFund, err := sdk.ParseCoinsNormalized("0umed")
+	if err != nil {
+		return nil, sdkerrors.Wrapf(err, "error in parsing coin")
+	}
+
+	_, err = k.wasmKeeper.Execute(ctx, nftContractAcc, redeemerAcc, transferNFTMsgBz, zeroFund)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	nftRedeemReceipt := types.NewDataAccessNFTRedeemReceipt(redeemNFT.PoolId, redeemNFT.Round, redeemNFT.NftId, uint64(ctx.BlockHeight()), redeemNFT.Redeemer)
+
+	err = k.SetDataPassRedeemReceipt(ctx, *nftRedeemReceipt)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	return nftRedeemReceipt, nil
+}
+
+func (k Keeper) GetDataPassRedeemReceipt(ctx sdk.Context, poolID, round uint64, redeemer sdk.AccAddress) (types.DataPassRedeemReceipt, error) {
+	key := types.GetKeyPrefixNFTRedeemReceipt(poolID, round, redeemer)
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(key)
+
+	var receipt types.DataPassRedeemReceipt
+	err := k.cdc.UnmarshalBinaryLengthPrefixed(bz, &receipt)
+	if err != nil {
+		return types.DataPassRedeemReceipt{}, sdkerrors.Wrap(types.ErrGetDataPassRedeemReceipt, err.Error())
+	}
+
+	return receipt, nil
+}
+
+func (k Keeper) SetDataPassRedeemReceipt(ctx sdk.Context, redeemReceipt types.DataPassRedeemReceipt) error {
+	redeemer, err := sdk.AccAddressFromBech32(redeemReceipt.Redeemer)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, err.Error())
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	receiptKey := types.GetKeyPrefixNFTRedeemReceipt(redeemReceipt.PoolId, redeemReceipt.Round, redeemer)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(&redeemReceipt)
+	store.Set(receiptKey, bz)
+
+	return nil
+}
+
+func (k Keeper) GetRedeemerDataPass(ctx sdk.Context, poolID uint64, redeemer sdk.AccAddress) ([]string, error) {
```

Comment:
Yes, my mistake.
I'll change like [this](https://github.com/medibloc/panacea-core/blob/5db2d0e4ed7dad3593505bb4c7e8b17d2cd9d712/x/datapool/keeper/pool.go#L634) in next PR. How about it?

Reply:
Well, I'm not sure what the difference is between the link you gave me and the current source. 🤔

Reply:
Ah, sorry. I was confused the branch. I think your idea is more better. The `GetPool` will be called over and over again. I understand. Thanks!

Reply:
> Does @hansol-medi mean taking all the DataPass that the redeemer owns?

No I didn't mean that. This function is for getting a list of DataPassRedeemReceipts used by redeemer in a pool.

If this `GetRedeemerDataPass()` accepts contractAddress as an argument, the caller should always know the contractAddress, which may be retrieved from `GetPool()`. In other words, this function may have a dependency on the `GetPool()` function.

What if we need a query for getting a list of DataPassRedeemReceipts used by redeemer in a pool? Then, I think, the argument of poolID makes sense as an independent function.

Umm, however, if we need that query above, we can get pool contractAddress first, and then get redeem receipts and return it. I'm not sure which is the best way..

Reply:
Yes. I thought that it was unnecessary to search the pool again here because it had already looked up the pool once in this upper function.
There is nothing wrong with what @hansol-medi said.

If I could implement it, I think I could try something like this.
Of course, this is just my opinion, and @inchori  can go with the implementation as it is now.

First, since `nftContractAcc` has already been declared in the parent function, it is passed as a parameter like this.
```go
func (k Keeper) GetRedeemerDataPass(ctx sdk.Context, poolID uint64, nftContractAcc, redeemer sdk.AccAddress) ([]string, error) {
    // code...
}
```
If you need to call the GetRedeemer function only with poolID, you can add it as follows.
```go
func (k Keeper) GetRedeemerDataPass(ctx sdk.Context, poolID uint64, redeemer sdk.AccAddress) ([]string, error) {
    pool, err := k.GetPool(ctx, poolID)
    if err != nil {
	return nil, sdkerrors.Wrapf(types.ErrPoolNotFound, err.Error())
    }

    acc, err := sdk.AccAddressFromBech32(pool.NftContractAddr)
    if err != nil {
    	return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, err.Error())
    }
    return k.GetRedeemerDataPassWithNftContractAcc(ctx, poolID, acc, redeemer)
}

func (k Keeper) GetRedeemerDataPassWithNftContractAcc(ctx sdk.Context, poolID uint64, nftContractAcc, redeemer sdk.AccAddress) ([]string, error) {
    // code...
}
```

Again, this is just my opinion, please keep in mind.

Reply:
Good. It's enough with the `GetRedeemerDataPassWithNftContractAcc()` for now, I think.
When we need `GetRedeemerDataPass()` or the related query, we can add it.

Reply:
Thanks for discussion. For now, I'll implement `GetRedeemerDataPassWithNftContractAcc()` and use it. But I'll leave `GetRedeemerDataPass()` method as well.